src/main/java/TaskManager.java

```diff
@@ -0,0 +1,190 @@
+        if(index> tasks.size() || index<=0){
+            DukeInvalidInputException e = new DukeInvalidInputException("Invalid input");
+            throw e;
+        }
+        if(markDone){//to mark it as done
+            tasks.get(index-1).setDone();
+            System.out.println("____________________________________________________________");
+            System.out.println("Nice! I've marked this task as done: ");
+        }
+        else{
+            tasks.get(index-1).setUndone();
+            System.out.println("____________________________________________________________");
+            System.out.println("OK, I've marked this task as not done yet:");
+        }
+        System.out.println(tasks.get(index-1));
+        System.out.println("____________________________________________________________");
+    }
+
+    /*
+    relatedToMark is a public method which checks if the user input is related to "mark" or "unmark" action.
+    If yes, it will call changeTaskStatus to do the corresponding task and return true. Otherwise, it returns false.
+    It is able to handle DukeInvalidException and NumberFormatException from changeTaskStatus method.
+     */
+    public boolean relatedToMark(String reply){
+        String[] words = reply.split(" ");
+        if (words[0].equalsIgnoreCase("mark") || words[0].equalsIgnoreCase("unmark")) {
+            if(words.length==1) {
+                System.out.println("OOPS!!! The mark/ unmark index should not be empty!");
+                return true;
+            }
+            try {
+                int index = Integer.parseInt(words[1]);
+                boolean markDone = words[0].equalsIgnoreCase("mark")? true: false;
+                changeTaskStatus(index, markDone);
+            } catch (DukeInvalidInputException | NumberFormatException e) {
+                System.out.println("OOPS!!! The mark index you've entered is invalid!");
+            }
+            finally {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /*
+    saveData is a public method to save current data into the hard disk (relative path: src\\DataSrc\\taskList.txt)
+     */
+    public void saveData() throws IOException {
+        FileWriter writer = new FileWriter(".\\src\\DataSrc\\taskList.txt");
+        for(Task task: tasks){
+            boolean taskDone = task.getStatusIcon().equalsIgnoreCase("X");
+            String taskType = task.toString().substring(1,2);
+            String taskInfo = taskType+","+taskDone+","+task.getDescription();
+            switch(taskType){
+                case "D":
+                    Deadline deadline = (Deadline) task;
+                    taskInfo = taskInfo + "," + deadline.getBy();
+                    break;
+                case "E":
+                    Event event = (Event) task;
+                    taskInfo = taskInfo + "," + event.getAt();
+                    break;
+            }
```

Comment:
There should be no indentation for the case clauses, to comply with the coding standard.